File: benchmarks/suites/syscalls.py
"""syscalls — strace-counted file operations behind `skit list --json` (the direct
per-entry-I/O evidence for the future summary-index PR) and the network-syscall count
expected to be zero on warm read-only paths. Linux + strace only; skips recorded.

The metric IDs are unsuffixed (`syscalls.list_json.*`) and defined at exactly one
library size — the plan carries a single N (1000) by design."""

from __future__ import annotations

import subprocess
import sys
from typing import TYPE_CHECKING

from ..parsers import FILE_OP_SYSCALLS, NETWORK_SYSCALLS, count_group, strace_counts
from ..results import Metric, SuiteOutput
from ._env import RunCtx, bench_env

if TYPE_CHECKING:
    from ..pipeline import SuitePlan


def run(ctx: RunCtx, plan: SuitePlan) -> SuiteOutput:
    if sys.platform != "linux":
        return SuiteOutput.skip_all("syscalls", "not Linux")
    if ctx.strace is None:
        return SuiteOutput.skip_all("syscalls", "strace not found")
    if len(plan.ns) != 1:
        raise RuntimeError("syscalls metric IDs are unsuffixed: the plan must carry one N")
    n = plan.ns[0]
    env = bench_env(ctx, ctx.datasets[n].root)
    table_file = ctx.workdir / f"strace_n{n}.txt"
    subprocess.run(  # noqa: S603 — fixed-shape strace argv
        [ctx.strace, "-f", "-c", "-o", str(table_file), ctx.skit, "list", "--json"],
        cwd=ctx.workdir,
        env=dict(env),
        check=True,
        stdout=subprocess.DEVNULL,
    )
    counts = strace_counts(table_file.read_text(encoding="utf-8"))
    output = SuiteOutput(suite="syscalls")
    output.metrics["syscalls.list_json.file_ops"] = Metric(
        value=float(count_group(counts, FILE_OP_SYSCALLS)), unit="count", n=1
    )
    output.metrics["syscalls.list_json.network"] = Metric(
        value=float(count_group(counts, NETWORK_SYSCALLS)), unit="count", n=1
    )
    output.raw[f"n{n}"] = counts
    return output

File: crates/skit-benchmarks/src/suites/syscalls.rs
//! Linux system-call census of the warm JSON list path.

use std::{collections::BTreeMap, fs};

use crate::{
    Metric, SuiteKind, SuiteOutput, SuitePlan,
    parsers::{FILE_OP_SYSCALLS, NETWORK_SYSCALLS, count_group, strace_counts},
    process::{ProcessSpec, run as run_process},
    runner::{PROBE_TIMEOUT, RunContext, path_arg},
};

use super::SuiteError;

pub(super) fn run(context: &RunContext, plan: &SuitePlan) -> Result<SuiteOutput, SuiteError> {
    #[cfg(not(target_os = "linux"))]
    {
        let _ = (context, plan);
        return Ok(SuiteOutput::skip_all(SuiteKind::Syscalls, "not Linux"));
    }
    #[cfg(target_os = "linux")]
    run_linux(context, plan)
}

#[cfg(target_os = "linux")]
fn run_linux(context: &RunContext, plan: &SuitePlan) -> Result<SuiteOutput, SuiteError> {
    let Some(strace) = &context.strace else {
        return Ok(SuiteOutput::skip_all(
            SuiteKind::Syscalls,
            "strace not found",
        ));
    };
    let [n] = plan.library_sizes.as_slice() else {
        return Err(SuiteError::Contract(
            "syscalls IDs are unsuffixed; plan must carry one N".to_owned(),
        ));
    };
    let table = context.workdir.join(format!("strace_n{n}.txt"));
    run_process(&ProcessSpec {
        argv: vec![
            path_arg(strace),
            "-f".to_owned(),
            "-c".to_owned(),
            "-o".to_owned(),
            table.display().to_string(),
            path_arg(&context.skit),
            "list".to_owned(),
            "--json".to_owned(),
        ],
        cwd: context.workdir.clone(),
        env: context.environment(*n)?,
        timeout: PROBE_TIMEOUT,
        check: true,
    })?;
    let text = fs::read_to_string(&table).map_err(|error| {
        SuiteError::Contract(format!("could not read {}: {error}", table.display()))
    })?;
    let counts = strace_counts(&text)?;
    Ok(SuiteOutput {
        suite: SuiteKind::Syscalls,
        duration_seconds: 0.0,
        metrics: BTreeMap::from([
            (
                "syscalls.list_json.file_ops".to_owned(),
                Metric::single(count_group(&counts, FILE_OP_SYSCALLS) as f64, "count"),
            ),
            (
                "syscalls.list_json.network".to_owned(),
                Metric::single(count_group(&counts, NETWORK_SYSCALLS) as f64, "count"),
            ),
        ]),
        skipped: Vec::new(),
        raw: BTreeMap::from([(
            format!("n{n}"),
            serde_json::to_value(counts).expect("counts serialize"),
        )]),
    })
}
